### PR TITLE
[WIP] Per sample gradients mechanism

### PR DESCRIPTION
**Pull request by samdow**
Adds a mechanism to compute efficient per sample gradients on modules.

The current intended way of using them is:
```python
model: torch.nn.Module
input: tensor[N, *]

with model.compute_per_sample_grad(batch_size=N):
  res = model(input)
  res.backward()

# Each parameter of model will now have a `grad_sample` field with the correct per sample grads.
# The `grad` field will be empty
```

TODO:
- [x] remove functions added to native_functions.yaml
- [ ] add documentation about best practices on using it
- [ ] add negative tests for non-supported sample inputs
~add support for non-first element batch size~ For now, this will only show up with SequenceBias, which is a custom Opacus module. We'll end up making this decision on a per function basis

---------------------
API options from discussion (11/9):
```python
# Opt 1
with element_wise_grad(model):
    model(inp).sum().backward()

    for p in model.parameter():
        print(p.grad_sample) # RuntimeError
        print(p.grad is not None) # has per-sample gradient

for p in model.parameter():
    print(p.grad_sample is not None)
    print(p.grad is None)



# Opt 2
out = nn.util.per_sample_call(model, inp)
out.sum().backward()

for p in model.parameter():
    print(p.grad_sample is not None)
    print(p.grad is None)


# Opt 3
some_dict = model.state_dict()
for k in some_dict:
    some_dict[k] = PerSampleObj(some_dict[k])

out = nn.util.stateless.functional_call(model, some_dict, inp)
out.sum().backward()

for k in some_dict:
    print(some_dict[k].grad)
for p in model.parameter():
    print(p.grad is None)
    
# Opt 4
model.per_sample_model_()
out = model(input)
out.sum().backward()

for p in model.parameter():
    print(p.grad_sample is not None)
    print(p.grad is None)
```

### CODE REVIEW

File: torch/expanded_weights/repeat_expanded_weights.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import torch
+
+# This acts differently because repeat is only a method so we need to alter ExpandedWeight
+class RepeatPerSampleGrad(torch.autograd.Function):
```
Comment (zou3519) [inline torch/expanded_weights/repeat_expanded_weights.py:4]: there's no `implements_per_sample_grads` here -- is this used?

File: torch/expanded_weights/layer_norm_expanded_weights.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+from torch.expanded_weights.expanded_weights_utils import\
+    forward_helper, grad_if_exists, grad_if_exists_for_input, sum_over_all_but_batch_and_last_n
+
+@implements_per_sample_grads(F.layer_norm)
+class LayerNormFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *expanded_args):
+        input = expanded_args[0]
+        normalized_shape = expanded_args[1]
+        assert len(input.shape) > len(normalized_shape)
```
Comment (zou3519) [inline torch/expanded_weights/layer_norm_expanded_weights.py:14]: A lot of these assertions (in other functions as well) should probably be RuntimeErrors so that they look nicer to the user
